Update README to provide installation instuctions

olokelo opened the pull request:
Added instructions how to compile aom, libavif and this module for Windows and Linux.
It may by useful for new users.
Also added `sample.avif` in examples folder for easier testing.

Julian commented:
Hi! This kind of PR is great, thanks very much for doing it!

CI is failing for some trivial reasons -- it's running [`pre-commit`](https://pre-commit.com/#usage) and likely just failing for trailing whitespace or the like. Let me know if it's unclear and will try to help. Also may be worth mentioning that in the README you're adding I suppose.

Would also slightly prefer sticking with rst rather than changing to markdown -- I usually have rst READMEs personally. It also is convenient if/when we have Sphinx documentation.

I have to read what you wrote carefully but a quick scan is like I said fantastic so thanks again for writing this up, it's something I obviously should have done myself but wanted to get the example passing and haven't had a chance this past week to get back to this library.

Obviously there's plenty more to do... Thanks again!

olokelo commented:
Hi and thanks for Your kind response!

I'm somewhat new to open source so I don't quite understand all those CI things, but from my understanding I forgot to change tests that are still looking for old README.rst file which I removed. As for rst and md I was kinda expecting that it might not be the best thing to change it. Unfortunately my markdown editor doesn't support rst but it shouldn't be a big deal to change it.

I have to say that I really like your idea of this project and in my opinion it's great! I wanted to do something like that myself for a while, but since You were first I decided just to contribute to Your project and I'm glad that it helped you. I want to collaborate on this project but first I should probably learn some cffi, because I never tried it before. I thought that maybe Pillow image plugin should be on the list things to do, but it's Your choice of course. For now I decided to do something like installation guide for this project, because I had some problems installing it.

I already see some typos in my guide so I will correct it as soon as possible. I should also do something with those tests and convert documents to rst format. I will try to do it 😄.

Julian commented:
> I'm somewhat new to open source so I don't quite understand all those CI things, but from my understanding I forgot to change tests that are still looking for old README.rst file which I removed. 

Cool, welcome! (Side note: I've considered it'd be nice if I added a GitHub label saying `new-to-open-source` so that folks who wanted to could use it -- I'm more than happy to explain anything in more detail than my usual curt responses I just never know who needs the extra explanation. Let me know whether you think you'd find such a thing useful if a project had it!)

> I forgot to change tests that are still looking for old README.rst file which I removed

So a good first thing to do is to [install pre-commit](https://pre-commit.com/#install) and run it locally via `pre-commit run --all-files` in the repository, so you can confirm locally what it says.

From looking at CI output, some of what it will be flagging will be trailing whitespace (i.e. lines with extra spaces left at the end of them). E.g. if you highlight [this line](https://github.com/Julian/avif/pull/1/files#diff-ebb830deebbb2a2b94b79414674034d17e33edf240e75ce16ed5762e45c6aaaeR78) in GitHub you'll see there are empty spaces on it. I think after you run pre-commit it'll just fix everything for you automatically, but I can have a closer look if you'd like at any errors you don't follow.

> I want to collaborate on this project but first I should probably learn some cffi, because I never tried it before.

CFFI is awesome yes! And your contributions will be super welcome!

Looking forward to your next update, let me know whenever it's ready to look at! And thanks again.

olokelo commented:
Hi, Your responses are really helpful. Thank You.

I managed to rewrite everything in rst again, fix some typos, install pre-commit and pass all checks  🎉.
Thanks again for Your awesome and detailed explanation!
I also tested my guide on a few more operating systems.

Sorry if I made a spam with those commits but at first github didn't like my rst formatting and added quotes everywhere in the installation guide so I had to change it.

I personally think that Your idea with GitHub label might be really useful for someone new to open source. For me at least it would be a useful thing. There are probably more programmers like me that started out by learning Python so they will make their first contributions to Python projects. I was somewhat nervous at first committing my work to other repositories because I was afraid of doing something wrong so maybe telling somebody that I'm new to it would make me feel better.

I feel that the changes are complete and ready to look at. Everything now looks and works just like it should. I just wanted to know if You have any suggestions.

Julian commented:
> I managed to rewrite everything in rst again, fix some typos, install pre-commit and pass all checks 🎉. Thanks again for Your awesome and detailed explanation! I also tested my guide on a few more operating systems.

Amazing, well done. I'm somewhat shot from the work week, but I'll be having a look at this tomorrow morning. I'll likely leave a few comments -- if at any point you'd rather I address them myself just let me know, otherwise I'll assume you're OK with making them (sometimes nitpicky comments can get annoying to address and you'd rather work on the next more interesting thing, which is totally fine).

> I personally think that Your idea with GitHub label might be really useful for someone new to open source. 

And thanks! Good to hear, I may give that a shot on some other projects of mine.

Julian commented on INSTALL.linux.rst:
```diff
@@ -0,0 +1,123 @@
+  sudo apt-get install cmake make ninja-build gcc g++ nasm git python3-dev python3-pip
+
+
+Build Steps
```
So, will leave specific comments below, but one general thing -- most of what's in these steps generally shouldn't be required for an end-user to install this library, so we should probably include the more "usual" install instructions. We're pushing wheels to PyPI: https://pypi.org/project/avif/#files which happens here: https://github.com/Julian/avif/blob/main/.github/workflows/packaging.yml so most users should not have to run any of the below.

Did you get something different when you installed purely via pip? I assume if you did, possibly because you wrote you're on ARM64, where yeah I didn't push ARM64 wheels -- I forget whether GH Actions offers ARM64 builders for Linux, I believe it actually does, and that if we enabled that you'd also get a wheel.

What you wrote up certainly is correct for a platform that doesn't currently have wheels being built for it, or to build a development build, then you'd go through the below to build from source, but yeah most users should just be able to run `pip install avif`. E.g. see below:

```
⊙  docker run --rm -it python:latest /bin/bash                                                                                                                          julian@Airm
root@32236ac1e544:/# python -m pip install avif
Collecting avif
  Downloading avif-0.3.0-cp39-cp39-manylinux2010_x86_64.whl (77 kB)
     |████████████████████████████████| 77 kB 2.4 MB/s
Collecting cffi>=1.0.0
  Downloading cffi-1.14.4-cp39-cp39-manylinux1_x86_64.whl (405 kB)
     |████████████████████████████████| 405 kB 11.7 MB/s
Collecting attrs
  Downloading attrs-20.3.0-py2.py3-none-any.whl (49 kB)
     |████████████████████████████████| 49 kB 5.8 MB/s
Collecting pycparser
  Downloading pycparser-2.20-py2.py3-none-any.whl (112 kB)
     |████████████████████████████████| 112 kB 8.1 MB/s
Installing collected packages: pycparser, cffi, attrs, avif
Successfully installed attrs-20.3.0 avif-0.3.0 cffi-1.14.4 pycparser-2.20
WARNING: You are using pip version 20.3.3; however, version 21.0 is available.
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
root@32236ac1e544:/# python -c 'import avif'
```
```

Julian commented on INSTALL.linux.rst:
```diff
@@ -0,0 +1,123 @@
+
+#. Install cffi module from pypi:
+
+   - ``sudo pip3 install --upgrade cffi``
```
You should avoid running `pip` with `sudo` (which will apply again below).

But specifically for `cffi`, `avif` declares it depends on CFFI already, so you shouldn't have to install it manually at all, it'll be installed automatically, *except* if you're documenting how to run rebuilding the shared library during development, i.e. running `python avif/_build.py`? Then yes you need a virtualenv where either avif has been installed or at least cffi is present.

Julian commented on INSTALL.linux.rst:
```diff
@@ -0,0 +1,123 @@
+
+   Edit the ``libavif/ext/aom.cmd`` script:
+
+   - Add ``-DCMAKE_INSTALL_PREFIX=/usr -DBUILD_SHARED_LIBS=1`` after ``cmake -G Ninja -DCMAKE_BUILD_TYPE=Release`` and keep rest of the line as is
```
I'd rather not tell users to install things into /usr/, especially on Linux -- `/usr/local` or `/opt` is generally where non-package-manager-managed things should go.

Julian commented on INSTALL.linux.rst:
```diff
@@ -0,0 +1,123 @@
+
+   - Ensure that you are in the main libavif directory (``echo $PWD`` should give ``~/python-avif/libavif``)
+   - ``cd cmake``
+   - ``cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DAVIF_CODEC_AOM=ON``
```
Same here on /usr

Julian commented on INSTALL.linux.rst:
```diff
@@ -0,0 +1,123 @@
+#. If build succeeded, you can finally install this module from python:
+
+   - ``cd ~/python-avif/avif``
+   - ``python setup.py build``
```
In modern Python packaging one should generally never run `python setup.py anything` -- here the way to install is `pip install .`, but again would rather not recommend anyone ever use `sudo pip`, so perhaps we should document how to install into a virtualenv.

Julian commented on INSTALL.linux.rst:
```diff
@@ -0,0 +1,123 @@
+Build Requirements
+------------------
+
+- cmake
```
I think it'd be nice to separate prerequisites for libavif from prerequisites for this library -- specifically for prerequisites / install instructions for libavif I'd love to link to upstream instructions rather than rewriting them. Those are more likely to stay up to date as libavif itself evovles.

Julian commented on INSTALL.win.rst:
```diff
@@ -0,0 +1,129 @@
+
+2. Install cffi module from pypi:
+
+   - ``pip install --upgrade cffi``
```
Same here on CFFI, this shouldn't require manual installs and comments from the Linux file.

Julian reviewed (changes requested):
Thanks again here! I left some comments (and actually pushed some immediate tiny commits because I notice you had Windows line endings in files, so I enabled pre-commit to warn about those and ensure all files have *nix line endings).

Let me know what you think!

olokelo commented:
Hi, thanks for Your detailed comments and review.

I will try to change most things that You requested but I also have some questions.

Firstly sorry for my line endings, I started writing on Windows and then switched to my Linux pc to test everything on virtual machines and I didn't pay attention to line endings.

As for pip installation I found this package on pypi. I tried it and everything worked fine (side note: I'm using amd64 architecture but probably releasing some arm wheels wouldn't be that bad idea especially since world right now will be slowly transitioning to arm thanks to Apple). The only problem for me was the lack of av1 decoder (maybe You know how to fix that):
```
$ ls -la /usr/local/lib | grep libaom
-rw-r--r--  1 root root    9776744 Jan 17 22:57 libaom.a
lrwxrwxrwx  1 root root         11 Jan 17 22:57 libaom.so -> libaom.so.2
lrwxrwxrwx  1 root root         15 Jan 17 22:57 libaom.so.2 -> libaom.so.2.0.0
-rw-r--r--  1 root root    7791136 Jan 17 22:55 libaom.so.2.0.0
$ export LD_LIBRARY_PATH=/usr/local/lib
$ pip install --user avif
Collecting avif
  Downloading avif-0.3.0-cp38-cp38-manylinux2010_x86_64.whl (77 kB)
     |████████████████████████████████| 77 kB 633 kB/s 
Requirement already satisfied: attrs in /usr/lib/python3/dist-packages (from avif) (19.3.0)
Requirement already satisfied: cffi>=1.0.0 in /usr/local/lib/python3.8/dist-packages/cffi-1.14.4-py3.8-linux-x86_64.egg (from avif) (1.14.4)
Requirement already satisfied: pycparser in /usr/local/lib/python3.8/dist-packages/pycparser-2.20-py3.8.egg (from cffi>=1.0.0->avif) (2.20)
Installing collected packages: avif
Successfully installed avif-0.3.0
$ python -c 'import avif'
$ python avif_example_decode_file.py sample.avif 
Traceback (most recent call last):
  File "avif_example_decode_file.py", line 34, in <module>
    main()
  File "avif_example_decode_file.py", line 17, in main
    image = decoder.parse_path(filename)
  File "/home/oloke/.local/lib/python3.8/site-packages/avif/_api.py", line 40, in parse_path
    _succeed(lib.avifDecoderParse(self._decoder))
  File "/home/oloke/.local/lib/python3.8/site-packages/avif/_api.py", line 19, in _succeed
    raise AVIFError(ffi.string(c_string).decode("utf-8"))
avif._api.AVIFError: No codec available
```

After that I decided to compile everything myself and then it worked. Also as far as I know, there are no wheels for Windows. I wanted to try it out on my Windows machine too so I compiled it from source (luckily libavif and libaom have good support for this platform).
I totally agree that we should inform users that they can install package from pypi. I will include it in readme before the actual compilation steps.


> In modern Python packaging one should generally never run `python setup.py anything` -- here the way to install is `pip install .`, but again would rather not recommend anyone ever use `sudo pip`, so perhaps we should document how to install into a virtualenv.

Oh I didn't know about this pip feature. Thanks, I will try to change it.


I will try to fix everything You pointed out and also I learned a lot from Your comments, thanks. I've been using Python and Linux for about 4 years but I'm still in high school right now so I'm not an expert by any meaning.
Thanks again for Your time and effort.

Julian commented:
> Firstly sorry for my line endings, I started writing on Windows and then switched to my Linux pc to test everything on virtual machines and I didn't pay attention to line endings.

All good! Never any need to apologize for these things, tests catch our mistakes :) which is why I added the pre-commit hook, now no one (myself included) can make the mistake.

> I will try to fix everything You pointed out and also I learned a lot from Your comments, thanks. I've been using Python and Linux for about 4 years but I'm still in high school right now so I'm not an expert by any meaning.

This is awesome, so well done. If you're this far along in high school you've got a good life in software ahead of you, so keep it up :)

> (side note: I'm using amd64 architecture but probably releasing some arm wheels wouldn't be that bad idea especially since world right now will be slowly transitioning to arm thanks to Apple).

Locally I'm actually on ARM myself, though the docker container I showed is x86-64 -- but yes I agree ARM wheels would be great to upload too, it just depends what's available in GitHub really (and in cibuildwheel).

> After that I decided to compile everything myself and then it worked.

I *think* this is more likely to do with being missing libavif than the wheel? But will check myself too.

olokelo commented:
Hi again, I tried to make changes to the instructions as you requested but I ran into some problems, so decided to ask you before committing.

First off, how do you want to separate prerequisites for libavif and Python module? Will something like this be ok?

Build Requirements
--------------------

- **Libavif:**
  - cmake
  - ninja
  - ...
- **Python module**:
  - python (3.7+)
  - pip
  - ...


Second thing is that I don't know how we can tell users to install `libavif` to location other than `/usr/`. Of course, it could be `/usr/local/` but then every time when user tries to import library, `LD_LIBRARY_PATH` must be set.
That's probably not the way to use bindings, because constantly using `LD_LIBRARY_PATH` is somewhat dangerous because of security reasons etc.

Then I tried to put the `libavif.so*` to site-packages (i saw that it's working like that when you install wheel from pypi) but no luck :(
The last thing I tried was to build wheel locally using `pip wheel .` and then `pip install avif-0.3.0*.whl`.
I think that should be fine but pip doesn't include `libavif.so` in wheel and I don't know how to do it.
When I downloaded official wheel from PyPI and unpacked it, I saw there is `avif.libs` folder and Python can find avif library when you try to import it. When I built wheel locally I got only `avif`, `avif-0.3.0.dist-info` and `_avif.abi3.so` and of course exception when importing module telling me that `libavif.so.9` cannot be found.
I also tried to create `avif.libs` folder manually and then put `libavif.so*` there but Python still was unable to find it.

Do you know how to install external library into site-packages or build the wheel with external library included in it? Or maybe you have any other idea to let Python find libavif when it's not in `/usr/lib/`?

Julian commented:
Will have to read your comment more carefully, but just on the last piece I see from skimming -- you can bundle e.g. `libavif.so`'s with the wheel via [auditwheel](https://github.com/pypa/auditwheel) -- doing so may be a decent idea but it comes with license implications (i.e. we have to be allowed to distribute libavif too.) I forget what license it's under, but we'd need to check.

olokelo commented:
Hi, sorry for my late answer. I've got rather tough week in school and I didn't have much time to work on changes.

However I managed to install auditwheel and add required libraries automatically to wheel built by pip.
I documented everything so that the user should have no problems building this package. Of course I also added note about PyPI in readme before actual compilation steps.

I also figured out that the dependency problem was about `_avif.abi3.so`. I was so desperated to sort of reverse engineer it just to check it dependencies and compare it with those in wheel from PyPI.
I've tried with `auditwheel` before but I haven't selected the appropriate plat (linux_x86_64). The default (manylinux) gave me an error complaining that my toolchain is too new and I got distracted about it. After I changed the plat everything worked just fine.

It *should* be good now. I think that everything you requested is now changed. Could you please review it?

In the meantime I also wrote a little Pillow plugin in Python to allow users open avif images using Pillow.
I have to test it a little bit more and I will probably commit it soon :)

Julian commented:
No worries! Quite busy here too so I understand.

> In the meantime I also wrote a little Pillow plugin in Python to allow users open avif images using Pillow. I have to test it a little bit more and I will probably commit it soon :)

This is awesome! That sounds great yeah.

I need to re-review this, probably tomorrow, but I'm also going to give you commit access to this repo :) Use it wisely (and yeah do loop me in on stuff, but you've got it now).

olokelo commented:
Hi, just finished writing a Pillow plugin for this library. Sorry if i committed into wrong branch, but I still don't know how it will work if I create a new one and this pull request still isn't merged. I could potentially merge it myself but it would be really bad to mess up something with your repo.

I tested my code and it worked fine in every case. My plugin supports animated and static avif files, it doesn't support HDR because as far as I know Pillow doesn't have image mode for it. Also I decided to add Pillow as requirement to install this module.

Thanks for giving me commit access to this repo but I think that for now I won't be using it. I just don't feel that confident to commit changes before your review. I also saw one failing check which I don't know what it is about so I will probably need some support.

And finally thanks for your support, you're awesome :)

Julian commented:
> Hi, just finished writing a Pillow plugin for this library. Sorry if i committed into wrong branch, but I still don't know how it will work if I create a new one and this pull request still isn't merged.

No worries at all there! That's mostly because of me being slow to merge this so apologies again about that, but what you've done here won't impact that, I'll probably first merge your initial work and then leave comments on the pillow stuff which looks great!

> Also I decided to add Pillow as requirement to install this module.

Nice! I think we may want to make it an optional dependency via a setuptools extra in case someone doesn't want to have to install pillow but will leave a comment on how to do that.

> Thanks for giving me commit access to this repo but I think that for now I won't be using it. I just don't feel that confident to commit changes before your review.

This is totally fine :) -- it's normal to be tentative on commit access so definitely feel free to keep doing exactly what you were doing so far. You'll get some more confidence over time anyhow (whether it's on this project or in general). But yeah wanted to give it to you nonetheless.

> I also saw one failing check which I don't know what it is about so I will probably need some support.

Will have a look and leave a tip.

And thanks again!

Julian commented:
Hello! I've merged your README changes, though there are still a few small tweaks I'm going to make.

Will have a look at the Pillow work soon!

Thanks again.

Julian commented:
GitHub seems a bit confused about the merge -- perhaps you might open another PR for the Pillow work from the same branch?

olokelo commented:
Thank you for your review!

I hope I helped you with the installation docs somehow.
I will try to open this new PR now :)

olokelo commented:
Ok, one more thing. It might be pretty basic question but I don't want to break something.
If I have my repo locally and it's behind your branch now, should I do `git pull upstream main` now or new PR could be done without it?
I ask because it wants me to merge your branch into mine and treat it as a commit and I don't know if it's right.

Julian commented:
> If I have my repo locally and it's behind your branch now, should I do git pull upstream main now or new PR could be done without it?
I ask because it wants me to merge your branch into mine and treat it as a commit and I don't know if it's right.

Yeah that should be fine! You can let it merge the upstream branch and then the new PR should just have the new changes.
